### PR TITLE
Implement error codes for the `mtree_verify` instructions

### PR DESCRIPTION
**Pull request by Overcastan**
This PR implements the ability to specify error codes for mtree_verify instruction the same way it is done for assert instructions.

TODO:
- [x] update changelog

Related issue: https://github.com/0xPolygonMiden/miden-vm/issues/1264

### CODE REVIEW

File: assembly/src/parser/grammar.lalrpop
Comment (bitwalker) [inline assembly/src/parser/grammar.lalrpop:400]: *Nit:* I would name this something a bit more precise, perhaps `InstWithErrorCode`? Most other instances of `Maybe*`  are macros which can be applied to a non-terminal, but that isn't the case here - the others are more precise about what is being parsed, so less ambiguous at first glance.

**Review by bitwalker — APPROVED** [review]
Looks good! Just one small nit, but your call on that one.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!

One outstanding thing, probably for a different PR, is figuring out how we can treat `MerklePathVerificationFailed` error in a similar way as `FailedAssertion`. Specifically, for `FailedAssertion`, we have a dedicated host handler which can process assertion errors in a custom way (i.e., by mapping them to user-friendly error messages). Should we do something similar for the `MerklePathVerificationFailed`?

If we do add a special handler to the `Host` for `MerklePathVerificationFailed` - should it cover only `mtree_verify` instruction, or other Merkle-tree related instructions as well?

As I mentioned, this is probably for a different PR - but I think we should have a good understanding of what we want to do before merging this PR.

**Comment by Overcastan** [comment]
We can add a Host handler for Merkle tree errors exactly the same as we do it for `Assert` now, with handler like `on_merkle_tree_error`. I think it makes sense to use it for all Merkle-tree related instructions: for now we have only `mtree_verify`, but probably other instructions eventually will have error codes as well.

By the way, I noticed that we call `on_assert_failed` Host function only for `Assert(err)` operation, but not for `U32assert2(err)`. Formally it is also an assert operation, although it is used in a slightly different way. Should we leave it as it is?

**Comment by bobbinth** [comment]
> By the way, I noticed that we call `on_assert_failed` Host function only for `Assert(err)` operation, but not for `U32assert2(err)`. Formally it is also an assert operation, although it is used in a slightly different way. Should we leave it as it is?

Ah - I didn't realize this was happening. Let's fix this (in a different PR). The one question would be is whether we can re-use the same function `on_assert_failed()` since now we need to return a different error variant, I believe. Or would we need to add another handler to the host.

In my mind, adding another function to the host is potentially too much because we'd end up with something like `on_assert_failed`, `on_u32_assert_failed`, `on_merkle_tree_error`. Maybe we have a single function in the host and specify the type of the assertion via an additional enum parameter (and this would include regular `assert`, `u32assert2`, and `mtree_verify`).

Let's try to solve this in a follow up PR.

**Pull request merged**
